### ISSUE TITLE
(V)AEGMM batch computation, instabilities and others.

### ISSUE THREAD

**Issue by RobertSamoilescu** [SeldonIO/alibi-detect#554]
Some issues to be addressed for (V)AEGMM:

- [ ] The current implementation of (V)AEGMM does not support batch computation when estimating the GMM components after the training procedure has finished. For AEGMM the problematic lines are [here](https://github.com/SeldonIO/alibi-detect/blob/6e4f5e0c86e562d0a92bb48d0a9690de0606a37b/alibi_detect/od/aegmm.py#L127-L128). Note that the current implementation feeds the entire dataset at once to compute the embedding representation and the mixture components. For large datasets and complex models this approach would not be possible as we would get OOM. Furthermore, the computation of the mean and covariance matrix is even more memory demanding. 
- [ ] The computation of the GMM parameters performs Cholesky decompostion of the covariance matrix. Note that in order to perform the Cholesky decomposition, the matrix has to be positive definite. The common trick is to add a diagonal matrix as [here](https://github.com/SeldonIO/alibi-detect/blob/6e4f5e0c86e562d0a92bb48d0a9690de0606a37b/alibi_detect/models/tensorflow/gmm.py#L45-L46). The problem with the current implementation is that the value of `eps` is hard-coded and in some cases it might not be enough to transform the matrix in a positive definite matrix. Thus, we need to expose this parameter to (V)AEGMM fitting method.
- [ ] Another numerical instability can arise from the regularization of the diagonal of the covariance matrix. The current implementation regularizes the entire diagonal, including the last (two) entries which come form the reconstruction error appended to the embedding representation of the autoencoder/compression network. In fact, in the paper there is a sentence which says "To avoid this issue, we penalize small values on the diagonal entries by ..., where d is the number of dimensions in the low-dimensional representations provided by the compression network." . The effect of the regularization is to increase the distance between the input and the reconstruction which is in conflict with the reconstruction loss, which in some cases can lead to training instabilities. Thus, we may need to remove the last entries from the regularization. Probably worth checking if [this implementation](https://github.com/tnakae/DAGMM)  addresses this issue. 
- [ ] We probably need to change this [reconstruction loss](https://github.com/SeldonIO/alibi-detect/blob/6e4f5e0c86e562d0a92bb48d0a9690de0606a37b/alibi_detect/models/tensorflow/losses.py#L77) (i.e. averaging over all dimension) with summation over features and averaging over instances. Probably worth checking [this implementation](https://github.com/tnakae/DAGMM/blob/3de2e9ab3a6ce7960df4c61809bdc78aa88da5f0/dagmm/compression_net.py#L117). With the current setting, the energy term can dominate the reconstruction which can lead to poor reconstruction.
